presto-kafka/src/main/java/io/prestosql/plugin/kafka/encoder/json/JsonRowEncoder.java

```diff
@@ -0,0 +1,267 @@
+            return dateFormatFunction(dataFormat, formatHint);
+        }
+        else if (type == TIME) {
+            return timeFormatFunction(dataFormat, formatHint);
+        }
+        else if (type == TIME_WITH_TIME_ZONE) {
+            return timeWithTZFormatFunction(dataFormat, formatHint);
+        }
+        else if (type instanceof TimestampType) {
+            return timestampFormatFunction(dataFormat, formatHint);
+        }
+        else if (type instanceof TimestampWithTimeZoneType) {
+            return timestampWithTZFormatFunction(dataFormat, formatHint);
+        }
+        else {
+            throw new IllegalArgumentException(format("Invalid type '%s' for temporal column '%s'", type.getDisplayName(), dataFormat, name));
+        }
+    }
+
+    private String currentColumnName()
+    {
+        return columnHandles.get(currentColumnIndex).getName();
+    }
+
+    @Override
+    protected void appendNullValue()
+    {
+        node.putNull(currentColumnName());
+    }
+
+    @Override
+    protected void appendLong(long value)
+    {
+        node.put(currentColumnName(), value);
+    }
+
+    @Override
+    protected void appendInt(int value)
+    {
+        node.put(currentColumnName(), value);
+    }
+
+    @Override
+    protected void appendShort(short value)
+    {
+        node.put(currentColumnName(), value);
+    }
+
+    @Override
+    protected void appendByte(byte value)
+    {
+        node.put(currentColumnName(), value);
+    }
+
+    @Override
+    protected void appendDouble(double value)
+    {
+        node.put(currentColumnName(), value);
+    }
+
+    @Override
+    protected void appendFloat(float value)
+    {
+        node.put(currentColumnName(), value);
+    }
+
+    @Override
+    protected void appendBoolean(boolean value)
+    {
+        node.put(currentColumnName(), value);
+    }
+
+    @Override
+    protected void appendString(String value)
+    {
+        node.put(currentColumnName(), value);
+    }
+
+    @Override
+    protected void appendByteBuffer(ByteBuffer value)
+    {
+        node.put(currentColumnName(), value.array());
+    }
+
+    @Override
+    protected void appendSqlDate(SqlDate value)
+    {
+        node.put(currentColumnName(), ((JsonDateTimeFormatFunction) formatFunctions.get(currentColumnIndex)).format(daysToMillis(value.getDays())));
+    }
+
+    @Override
+    protected void appendSqlTime(SqlTime value)
+    {
+        node.put(currentColumnName(), ((JsonDateTimeFormatFunction) formatFunctions.get(currentColumnIndex)).format(value.getMillis()));
+    }
+
+    @Override
+    protected void appendSqlTimeWithTimeZone(SqlTimeWithTimeZone value)
+    {
+        node.put(currentColumnName(), ((JsonDateTimeFormatFunction) formatFunctions.get(currentColumnIndex)).format(packDateTimeWithZone(value.getMillisUtc(), value.getTimeZoneKey())));
+    }
+
+    @Override
+    protected void appendSqlTimestamp(SqlTimestamp value)
+    {
+        if (session.isLegacyTimestamp()) {
```

Comment:
This looks wrong. It does not seem to me that you should not use same formatting function and just pass different version to it based on selected timestamp semantic.
As the semantics is different the formatting should be different. I.e it does not make sense to append any timezone information for non-legacy semantics in output string.

It is hard for me to wrap my head around it though.

Reply:
Ik the timestamp changes have me confused as well (clearly hahah). I think I've been testing with the legacy timestamp, which would explain why I haven't encountered a problem with this yet, so I'll add different functions and test with the new timestamp.

Reply:
It seems to me that cleanest approach would be to not allow encoding timestamp in non-legacy semantic to any format which clearly denotes an instant (i.e. explicitly specifies a timezone in a value). The problem with that is it would be not compatible with what we already have in decoders (which do not care about timestamp semantics whatsoever). 

Maybe we should assume that logic in decoders only makes sense for legacy timestamp semantics, and enforce that with check in the connector. Then we can focus on implementing just legacy semantics of timestamp in encoders in a way that is compatible with decoders.

Then we can add non-legacy semantics to both decoders and encoders?

The problem with that is that we would break people which use kafka timestamp decoders with timestamp semantic set to non-legacy. I do not expect that it would be many users, but one never knows.

@findepi, as an expert on hard issues, can you chime in here?

Reply:
@losipiuk i like your proposal.
i would only add that there seems to be no fundamental problem in decoding `timestamp` (without zone) from a value containing a zone. You could think of it as a CAST.
Eg JDBC ResultSet.getDate can be called on a date or timestamp, and returns a date part of the value.

I am OK with dropping this though, for non-legacy. We can always re-add this if we learn about particular requirements.

Reply:
Thanks for comment.

> @losipiuk i like your proposal.
> i would only add that there seems to be no fundamental problem in decoding `timestamp` (without zone) from a value containing a zone. You could think of it as a CAST.
> Eg JDBC ResultSet.getDate can be called on a date or timestamp, and returns a date part of the value.
> 
> I am OK with dropping this though, for non-legacy. We can always re-add this if we learn about particular requirements.

Yeah - it makes sense for decoding. Less so for encoding, as it not obvious what timezone use in encoded value. And even if timezone is statically defined by encoding format it is kinda smelly.